0867/Original-Submission/Src/quadlog-gampsi-slim/man/qgjqf.man
.TH QGJQF 3 "29 April 2000" "Version 1.00"
.\" WARNING: This file was produced automatically from file jacobi/qgjqf.f
.\" by fortran-to-man-page.awk on Tue Nov  4 07:39:18 MST 2003.
.\" Any manual changes will be lost if this file is regenerated!
.SH NAME
qgjqf \- Gauss-Jacobi logarithmic Quadrature with Function values
.\"=====================================================================
.SH SYNOPSIS
Fortran (77, 90, 95, HPF):
.RS
.B f77
.I "[ flags ] file(s) .\|.\|. -L/usr/local/lib -lgjl"
.RS
.nf
.B "SUBROUTINE qgjqf(x, w, y, z, alpha, beta, nquad, ierr)"
.B "INTEGER             ierr,        nquad"
.B "REAL*16             alpha,       beta,        w(*),        x(*)"
.B "REAL*16             y(*),        z(*)"
.fi
.RE
.RE
C (K&R, 89, 99), C++ (98):
.RS
.B cc
.I "[ flags ] -I/usr/local/include file(s) .\|.\|. -L/usr/local/lib -lgjl"
.br
Use
.RS
.B "#include <gjl.h>"
.RE
to get this prototype:
.RS
.nf
.B "void qgjqf(fortran_quadruple_precision x_[],"
.B "           fortran_quadruple_precision w_[],"
.B "           fortran_quadruple_precision y_[],"
.B "           fortran_quadruple_precision z_[],"
.B "           const fortran_quadruple_precision * alpha_,"
.B "           const fortran_quadruple_precision * beta_,"
.B "           const fortran_integer * nquad_,"
.B "           fortran_integer * ierr_);"
.fi
.RE
.RE
.PP
NB: The definition of C/C++ data types
.B fortran_
.IR xxx ,
and the mapping of Fortran external names to C/C++ external names,
is handled by the C/C++ header file.  That way, the same function
or subroutine name can be used in C, C++, and Fortran code,
independent of compiler conventions for mangling of external
names in these programming languages.
.\"=====================================================================
.SH DESCRIPTION
.PP
Compute the nodes and weights for the evaluation of the integral
.PP
.RS
.nf
\eint_{\(mi1}^{1} (1\(mix)^\ealpha(1+x)^\ebeta\eln(1+x)f(x)dx
            (\ealpha > \(mi1, \ebeta > \(mi1)
.fi
.RE
.PP
as the quadrature sum
.PP
.RS
.nf
\esum_{i=1}^N[W_i(\ealpha,\ebeta) (\eln 2) f(x_i(\ealpha,\ebeta)) \(mi Z_i(\ealpha,\ebeta)f(y_i(\ealpha,\ebeta))]
.fi
.RE
.PP
The nonlogarithmic ordinary Gauss-Jacobi integral
.PP
.RS
.nf
\eint_{\(mi1}^{1} (1\(mix)^\ealpha(1+x)^\ebeta f(x)dx
            (\ealpha > \(mi1, \ebeta > \(mi1)
.fi
.RE
.PP
can be computed from the quadrature sum
.PP
.RS
.nf
\esum_{i=1}^N[W_i(\ealpha,\ebeta) f(x_i(\ealpha,\ebeta)]
.fi
.RE
.PP
The quadrature is exact to machine precision for f(x) of
polynomial order less than or equal to 2*\fBnquad\fP\& \(mi 1.
.PP
This form of the quadrature requires only values of the
function, at 2*\fBnquad\fP\& points. For a faster, and slightly more
accurate, quadrature that requires values of the function and
its derivative at \fBnquad\fP\& points, see the companion routine,
qgjqfd().
.PP
On entry:
.PP
.RS
.TP \w'\fBalpha\fP\&'u+2n
\fBalpha\fP\& 
Power of (1\(mix) in the integrand (\fBalpha\fP\& > \(mi1).
.TP
\fBbeta\fP\& 
Power of (1+x) in the integrand (\fBbeta\fP\& > \(mi1).
.TP
\fBnquad\fP\& 
Number of quadrature points to compute. It
must be less than the limit MAXPTS defined
in the header file, 
.I maxpts.inc.
The default value chosen there should be large
enough for any realistic application.
.RE
.PP
On return:
.PP
.RS
.TP \w'\fBx\fP\&(1.\|.\fBnquad\fP\&)'u+2n
\fBx\fP\&(1.\|.\fBnquad\fP\&) 
Nodes of the Jacobi quadrature, denoted
x_i(\ealpha,\ebeta) above.
.TP
\fBw\fP\&(1.\|.\fBnquad\fP\&) 
Weights of the Jacobi quadrature, denoted
W_i(\ealpha,\ebeta) above.
.TP
\fBy\fP\&(1.\|.\fBnquad\fP\&) 
Nodes of the quadrature for \(mi(1\(mix)^\ealpha *
(1+x)^\ebeta ln((1+x)/2), denoted
y_i(\ealpha,\ebeta) above.
.TP
\fBz\fP\&(1.\|.\fBnquad\fP\&) Weights of the quadrature for
\(mi(1\(mix)^\ealpha * (1+x)^\ebeta ln((1+x)/2),
denoted Z_i(\ealpha,\ebeta) above.
.TP
\fBierr\fP\& 
Error indicator:
.nf
= 0 (success),
1 (eigensolution could not be obtained),
2 (destructive overflow),
3 (\fBnquad\fP\& out of range),
4 (\fBalpha\fP\& out of range),
5 (\fBbeta\fP\& out of range).
.fi
.RE
.PP
The logarithmic integral can then be computed by code like this:
.PP
.RS
.nf
\fC      qlgtwo = dlog(2.0q+00)
      sum = 0.0q+00
      do 10 i = 1,nquad
          sum = sum + qlgtwo*w(i)*f(x(i)) - z(i)*f(y(i))
   10 continue\fP
.fi
.RE
.PP
The nonlogarithmic integral can be computed by:
.PP
.RS
.nf
\fC      sum = 0.0q+00
      do 20 i = 1,nquad
          sum = sum + w(i)*f(x(i))
   20 continue\fP
.fi
.RE
.PP
.\"=====================================================================
.SH "SEE ALSO"
.BR qgjqfd (3),
.BR qgjqrc (3).
.\"=====================================================================
.SH AUTHORS
The algorithms and code are described in detail in
the paper
.RS
.I "Fast Gaussian Quadrature for Two Classes of Logarithmic Weight Functions"
.RE
in ACM Transactions on Mathematical Software,
Volume ??, Number ??, Pages ????--???? and
????--????, 20xx, by
.RS
.nf
Nelson H. F. Beebe
Center for Scientific Computing
University of Utah
Department of Mathematics, 110 LCB
155 S 1400 E RM 233
Salt Lake City, UT 84112-0090
Tel: +1 801 581 5254
FAX: +1 801 581 4148
Email: \fCbeebe@math.utah.edu\fP, \fCbeebe@acm.org\fP, \fCbeebe@computer.org\fP
WWW URL: \fChttp://www.math.utah.edu/~beebe\fP
.fi
.RE
and
.RS
.nf
James S. Ball
University of Utah
Department of Physics
Salt Lake City, UT 84112-0830
USA
Tel: +1 801 581 8397
FAX: +1 801 581 6256
Email: \fCball@physics.utah.edu\fP
WWW URL: \fChttp://www.physics.utah.edu/people/faculty/ball.html\fP
.fi
.RE
.\"==============================[The End]==============================
.\"=====================================================================
.\" This is for GNU Emacs file-specific customization:
.\" Local Variables:
.\" fill-column: 50
.\" End:


Changes
April 22 2021
   * Added 1013-1014 from Volume 47, number 1, (January 2021)
   * Added 1012 from Volume 46, number 4, (November 2020)
October 19 2020
   * Added 1011 from Volume 46, number 3, (September 2020)
August 7 2020
   * Added 1003-1006 from Volume 46, number 1, (April 2020)
   * Added 1007-1010 from Volume 46, number 2, (June 2020)
January 12 2020
   * Updated 999 at the request of the author -- updated user manual supplied.
January 7 2020
   * Added 999-1002 from Volume 45, number 4, (December 2019)
October 13 2019
   * Added 995-998 from Volume 45, number 3, (August 2019)
August 8 2019
   * Added 993 and 994 from Volume 45, number 2, (June 2019)
August 7 2019
   * Added 991 and 992 from Volume 45, number 1, (February 2019)
February 1 2019
   * Added 989 and 990 from Volume 44, number 3, (August 2018)
May 8 2018
   * Added 987 and 988 from Volume 44, number 3, (April 2018)
October 13 2017
   * Added 986 from Volume 44, number 2, (October 2017)
   * Corrected publication month for Algs 979-985 (now Oct).
September 30 2017
   * Added 979-985 from Volume 44, number 2, (September 2017)
   * Added 975-978 from Volume 44, number 1, (July 2017)
April 21 2017
   * Added 973, 974 and Remark on 936 from Volume 43, number 4 
     (March 2017)
February 3 2017
   * Added 968-972 from Volume 43, number 3 (January 2017)
September 16 2016
   * Added 965 from Volume 43, number 1 (August 2016)
   * Added 966, 967 from Volume 43, number 2 (September 2016)
   * Removed files 620 and 620.bib (all other algorithm files
     are only available as compressed files.
   * Replaced index.gz and index.html.gz by .zip files
September 14 2016:
   * Added 100.zip and 101.zip. These files are identical as
   the two algorithms are inter-related and the driver program
   uses both codes.
August 30 2016:
   * Added 963-964 from Volume 42, number 4 (July 2016)
July 23 2016
   * Updated contact information on CALGO web pages.
June 19 2016:
   * Added 960-962 from Volume 42, number 3 (June 2016)
   * Updated 916 to include the code submitted with Remark 
     in Volume 42, number 3 (Jun 2016)
June 15 2016:
   * Added 957 from Volume 42, number 1 (February 2016)
   * Added 958-959 from Volume 42, number 2 (June 2016)
May 30 2016
   * Added 955-956 from Volume 42, number 1 (February 2016)
May 27 2016:
   * Added new code supplied as Remark to Algorithm 673
     appearing in Volume 42, number 1.
December 22 2015:
   * Added 947-954 from Volume 41, numbers (1-4).
August 29 2015:
   * Mended a number of odd/broken compressed files that were on
     the CALGO site:
       755.gz, 756.gz, 823.gz replaced by zip files
       709.gz, 709a.gz, 709b.gz replaced by 709.zip
       573.gz, 573d.gz, 573s.gz replaced by 573.zip
       358.gz, 805.gz, 830.gz, 832.zip all updated
       620.zip added -- this contains all the 620 components.
September 26 2014
   * Updated 897 following accepted remark.
September 1 2014
   * Added 943-946 from Volume 40, number 4 (June 2014)
May 17 2014
   * Fixed problem with Algorithm 917 missing from indexes.
May 5 2014
   * Added 935-938 from Volume 40, number 2 (February 2014)
   * Added 939-942 from Volume 40, number 3 (April 2014)
October 9 2013
   * Added 929, 930 from volume 39, number 4 (July 2013)
   * Added 931-934 from volume 40, number 1 (October 2013)
July 18 2013
   * Added 928 from volume 39, number 3 (May 2013)
   * Updated 905 from code supplied by author thackerw@winthrop.edu
March 3 2013
   * Added 926-927 from volume 39, number 2 (February 2013)
January 25 2013
   * Replaced 923 after editing Licence file -- author request
January 3 2013
   * Added 924-925 from volume 39, number 1 (November 2012)
September 10 2012
   * Added 920-923 from volume 38, number 4 (August 2012)
June 17 2012
   * Added 917-919 from volume 38, number 3 (April 2012)
March 14 2012
   * Corrected typo in index files (Alg 916)
February 29 2012
   * Replaced 916 -- old version of code uploaded in error.
February  5 2012
   * Added 916 from Volume 38, number 2 (December 2012)
December 30 2011
   * Replaced 905 due to corrupt blas.f file in the original (pointed
     out by Ronald.M.Bloom@aero.org).
December 7 2011
   * Added 913-915 from Volume 38, number 1 (November 2011)
   * Updates 778 as per Remark in Volume 38, No. 1
April 12 2011
   * Added 909-911 from volume 37, number 4 (April 2011)
October 24 2011
   * Changed Algorithm 419 in line with changes from Remark on Algorithm 419 by
     David H. Withers CACM (March 1974) Vol 17 No 3 p. 157
     Pointed out by j.a.horsmeier@wanadoo.nl (Jos)
October 11 2010
   * Added 904-908 from volume 37, number 3 (September 2010)
May  20 2010
   * Added 902, 903 from volume 37, number 2 (April 2010)
Feb  27 2010
   * Added 898, 899, 900, 901 from volume 37, number 1 (January 2010)
July 29 2009
   * Added 895, 896, 897 from volume 36, number 3 (July 2009)
Apr  6 2009
   * Added 894 from volume 36, number 2 (March 2009)
Mar  30 2009
   * Added 891, 892, 893 from volume 36, number 1 (March 2009)
Feb  12 2009
   * Added 890 from volume 35, number 4 (February 2009)
Jan   6 2009
   * Added 885-889 from volume 35, number 3 (October 2008)
Sept  4 2008
   * Added 881-884 from volume 35, number 2 (July 2008)
July 25 2008
   * Added 876-877 from volume 34, number 4 (July 2008)
   * Added 878-880 from volume 35, number 1 (July 2008)
June 11 2008
   * Added 873-875 from volume 34, number 2 and 3 (March and May 2008)
Feb   8 2008
   * Added 867-869 from volume 33, numbers 3 and 4 (August 2007)
   * Added 870-872 from volume 34, number 1 (January 2008)
Aug   3 2007
   * Added 863-866 from volume 33, numbers 1 and 2.
July  3 2007
   * Updated 620 to include GAMS classifications to algorithms 814-862 inc.
Jan  10 2007
   * Updated 644 as a result of an accepted remark by Kodama
Jan   5 2007
   * Added 857-862 from volume 32, 4 (December 2006)
Nov  21 2006
   * Added 358 -- donated by John Burkardt <burkardt@csit.fsu.edu>.
Oct   5 2006
   * Added 854 from volume 32, 2 (June 2006)

0867/Src/quadlog-gampsi/jacobi/qgjqf.f
      SUBROUTINE  qgjqf(x, w, y, z, alpha, beta, nquad, ierr)
************************************************************************
*     (Gauss-Jacobi Quadrature with Function values)
*
*     Compute the nodes and weights for the evaluation of the integral
*
*         \int_{-1}^{1} (1-x)^\alpha(1+x)^\beta\ln(1+x)f(x)dx
*                                 (\alpha > -1, \beta > -1)
*
*     as the quadrature sum
*
*         \sum_{i=1}^N[W_i(\alpha,\beta) (\ln 2) f(x_i(\alpha,\beta))
*                      - Z_i(\alpha,\beta)f(y_i(\alpha,\beta))]
*
*     The nonlogarithmic ordinary Gauss-Jacobi integral
*
*         \int_{-1}^{1} (1-x)^\alpha(1+x)^\beta f(x)dx
*                                 (\alpha > -1, \beta > -1)
*
*     can be computed from the quadrature sum
*
*         \sum_{i=1}^N[W_i(\alpha,\beta) f(x_i(\alpha,\beta)]
*
*     The quadrature is exact to machine precision for f(x) of
*     polynomial order less than or equal to 2*nquad - 1.
*
*     This form of the quadrature requires only values of the
*     function, at 2*nquad points.  For a faster, and slightly more
*     accurate, quadrature that requires values of the function and
*     its derivative at nquad points, see the companion routine,
*     qgjqfd().
*
*     On entry:
*
*          alpha           Power of (1-x) in the integrand (alpha > -1).
*
*          beta            Power of (1+x) in the integrand (beta > -1).
*
*          nquad           Number of quadrature points to compute.  It
*                          must be less than the limit MAXPTS defined
*                          in the header file, maxpts.inc.  The default
*                          value chosen there should be large enough
*                          for any realistic application.
*
*     On return:
*
*          x(1..nquad)     Nodes of the Jacobi quadrature, denoted
*                          x_i(\alpha,\beta) above.
*
*          w(1..nquad)     Weights of the Jacobi quadrature, denoted
*                          W_i(\alpha,\beta) above.
*
*          y(1..nquad)     Nodes of the quadrature for
*                              -(1-x)^\alpha * (1+x)^\beta ln((1+x)/2),
*                          denoted y_i(\alpha,\beta) above.
*
*          z(1..nquad)     Weights of the quadrature for
*                              -(1-x)^\alpha * (1+x)^\beta ln((1+x)/2),
*                          denoted Z_i(\alpha,\beta) above.
*
*          ierr            Error indicator:
*                          = 0  (success),
*                            1  (eigensolution could not be obtained),
*                            2  (destructive overflow),
*                            3  (nquad out of range),
*                            4  (alpha out of range),
*                            5  (beta out of range).
*
*     The logarithmic integral can then be computed by code like this:
*
*          qlgtwo = dlog(2.0q+00)
*          sum = 0.0q+00
*          do 10 i = 1,nquad
*              sum = sum + qlgtwo*w(i)*f(x(i)) - z(i)*f(y(i))
*       10 continue
*
*     The nonlogarithmic integral can be computed by:
*
*          sum = 0.0q+00
*          do 20 i = 1,nquad
*              sum = sum + w(i)*f(x(i))
*       20 continue
*
*     [29-Apr-2000]
************************************************************************
*
*     External functions
*
      EXTERNAL            qvsum
*
      REAL*16             qsqrt,       qvsum
*
*     Parameter variables
*
      REAL*16    ONE
      PARAMETER           (ONE = 1.0Q+00)
      REAL*16    ZERO
      PARAMETER           (ZERO = 0.0Q+00)
*
      INCLUDE 'qlgtwo.inc'
      INCLUDE 'ecodes.inc'
      INCLUDE 'maxpts.inc'
*
*     Argument variables
*
      INTEGER             ierr,        nquad
*
      REAL*16             alpha,       beta,        w(*),        x(*)
      REAL*16             y(*),        z(*)
*
*     Local variables
*
      INTEGER             m,           n
*
      REAL*16             a(0:MAXPTS), b(0:MAXPTS), deltaw(MAXPTS)
      REAL*16             deltax(MAXPTS),           e(MAXPTS)
      REAL*16             p(MAXPTS),   pm(MAXPTS),  s(0:MAXPTS)
      REAL*16             t(0:MAXPTS)
*
*     Sanity checks on arguments:
*
      IF ((nquad .LE. 0) .OR. (nquad .GE. MAXPTS)) THEN
          ierr = enquad
          RETURN
      ELSE IF (alpha .LE. -ONE) THEN
          ierr = ealpha
          RETURN
      ELSE IF (beta .LE. -ONE) THEN
          ierr = ebeta
          RETURN
      ELSE
          ierr = eokay
      END IF
*
*     Compute recursion coefficients, nodes, and weights.
*
      CALL qgjqrc(a, b, s, t, alpha, beta, nquad, ierr)
      IF (ierr .NE. eokay) RETURN
*
*     Setup the tridiagonal Jacobi matrix whose elements are determined
*     by the a(*) and b(*) recursion coefficients.
*
      DO 100 n = 1, nquad
          y(n) = b(n-1)
          e(n) = qsqrt(a(n-1))
  100 CONTINUE
*
*     Diagonalize Jacobi matrix.  On return, y(*) contains the
*     quadrature nodes in ascending order, and e(*) is trashed.
*
      CALL qtql1(nquad, y, e, ierr)
      IF (ierr .NE. 0) THEN
          ierr = eeigen
          RETURN
      END IF
*
*     Compute the nodes and weights for order nquad for the first part
*     of the quadrature.
*
      CALL qgjqfd(x, w, deltaw, deltax, alpha, beta, nquad, ierr)
      IF (ierr .NE. eokay) RETURN
*
*     Normalize eigenvectors.
*
      DO 300 n = 1, nquad
          p(1) = ONE
          p(2) = (y(n) - b(0))*p(1)
          pm(1) = p(1)*(p(1)/t(0))
          if (nquad .GT. 1) pm(2) = p(2)*(p(2)/t(1))
          DO 200 m = 2, nquad - 1
              p(m + 1) = (y(n) - b(m - 1))*p(m) - a(m - 1)*p(m - 1)
              pm(m + 1) = p(m + 1)*(p(m + 1)/t(m))
  200     CONTINUE
          z(n) = ONE/qvsum(pm, nquad)
          IF (z(n) .EQ. ZERO) THEN
              ierr = eovflo
              RETURN
          END IF
  300 CONTINUE
*
      END
